### PR TITLE
MSI#856 extend salability interface

### PR DESCRIPTION
**Pull request by vadimjustus**
### Description
Change `\Magento\InventorySalesApi\Api\IsProductSalableInterface` to following
```php
    /**
     * Get is product in salable for given SKU in a given Stock
     *
     * @param string $sku
     * @param SalesChannelInterface $salesChannel
     * @return bool
     */
    public function execute(string $sku, SalesChannelInterface $salesChannel): bool;
```
*I decide not to add `$salesChannel` in addition to `$stockId` but replace `$stockId` completely, since it is always possible to resolve stock by given sales channel.*

Change `\Magento\InventorySalesApi\Api\IsProductSalableForRequestedQtyInterface` to following
```php
    /**
     * Get is product salable for given SKU in a given Stock for a certain Qty
     *
     * @param string $sku
     * @param SalesChannelInterface $salesChannel
     * @param float $requestedQty
     * @return ProductSalableResultInterface
     */
    public function execute(
        string $sku,
        SalesChannelInterface $salesChannel,
        float $requestedQty
    ): ProductSalableResultInterface;
```
*It is also necessary to extend `IsProductSalableForRequestedQtyInterface` interface in order to follow the same schema. In other case implementation end up with permanently switching between sales channel and stock id...*

I've introduce three additional classes in order to provide simple access to sales channels.
* `\Magento\InventorySales\Model\SalesChannelByWebsiteCodeProvider`
* `\Magento\InventorySales\Model\SalesChannelByWebsiteIdProvider`
* `\Magento\InventoryCatalog\Model\GetSalesChannelForCurrentWebsite`

### Fixed Issues (if relevant)

1. magento-engcom/msi#856: Need to extend Salability interface with SalesChannel

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/InventoryCatalog/Plugin/InventoryApi/SetDataToLegacyCatalogInventoryAtSourceItemsSavePlugin.php [hunk 1]
```diff
@@ -99,7 +103,7 @@ public function afterExecute(SourceItemsSaveInterface $subject, $result, array $
             $this->setDataToLegacyStockStatus->execute(
                 $sourceItem->getSku(),
                 (float)$sourceItem->getQuantity(),
-                (int)$this->isProductSalable->execute($sourceItem->getSku(), $this->defaultStockProvider->getId())
+                (int)$this->isProductSalable->execute($sourceItem->getSku(), $defaultSalesChannel)
```
Comment (maghamed) [inline app/code/Magento/InventoryCatalog/Plugin/InventoryApi/SetDataToLegacyCatalogInventoryAtSourceItemsSavePlugin.php]: this looks incorrect to me
```
    public function afterExecute(SourceItemsSaveInterface $subject, $result, array $sourceItems)
    {
        $defaultSalesChannel = $this->salesChannelByWebsiteIdProvider->execute(0);

       //... 

            $this->setDataToLegacyStockStatus->execute(
                $sourceItem->getSku(),
                (float)$sourceItem->getQuantity(),
                (int)$this->isProductSalable->execute($sourceItem->getSku(), $defaultSalesChannel)
            );
        }
    }
```

Before that logic looked as 
`(int)$this->isProductSalable->execute($sourceItem->getSku(), $this->defaultStockProvider->getId())`

and our business idea was to check whether Product is salable in Default Stock.

Currently `$defaultSalesChannel = $this->salesChannelByWebsiteIdProvider->execute(0);`
There is no guarantee that Main website would be assigned to Default stock.

Need to think whether we need to have two  Interfaces with `SalesChannel` and `Stock ` as parameters

**Review by maghamed — CHANGES_REQUESTED** [review]

**Comment by maghamed** [comment]
#904

**Comment by vadimjustus** [comment]
New update: I introduced two new Interfaces instead of `\Magento\InventorySalesApi\Api\IsProductSalableInterface`
* `\Magento\InventorySalesApi\Api\IsProductSalableForStockInterface`
* `\Magento\InventorySalesApi\Api\IsProductSalableForSalesChannelInterface`

@maghamed what do you think about this solution?

**Comment by maghamed** [comment]
Not an MSI bug, approved by PO

https://app.zenhub.com/workspace/o/magento-engcom/msi/issues/856#issuecomment-392030116

Product Enable/Disable should be figured out on the level of Configurable product.
Thus, having all the children as disabled - the parent product should get disabled as well

**Pull request closed without merging**